File: backend/apps/authentication/validators/email_validator.py
"""
EmailValidator — Chain of Responsibility node validating email format (SRP).
Extends BaseValidator (S1). Pattern: Chain of Responsibility. SOLID: SRP·OCP·LSP.
"""

from __future__ import annotations

import re

from core.base.base_validator import BaseValidator, ValidationResult

_EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")


class EmailValidator(BaseValidator):
    def validate(self, data: dict) -> ValidationResult:
        email = (data.get("email") or "").strip()
        if not _EMAIL_RE.match(email):
            return ValidationResult(
                is_valid=False,
                errors=["El correo electrónico no tiene un formato válido."],
                field_name="email",
            )
        return ValidationResult(is_valid=True)

File: frontend/src/modules/auth/validators/EmailValidator.ts
/**
 * EmailValidator — Chain of Responsibility node for email format (FE).
 * Extends BaseValidator (core). SOLID: SRP·OCP·LSP.
 */

import { BaseValidator, type ValidationResult } from '../../../core/base/BaseValidator'

// Clases de caracteres disjuntas ([^@\s.] nunca matchea el separador '.') →
// sin backtracking super-lineal (ReDoS) y sin puntos consecutivos en el dominio.
const EMAIL_RE = /^[^@\s]+@[^@\s.]+(?:\.[^@\s.]+)+$/

export class EmailValidator extends BaseValidator {
  validate(data: unknown): ValidationResult {
    const email = String((data as { email?: string })?.email ?? '').trim()
    if (!EMAIL_RE.test(email)) {
      return { isValid: false, errors: ['El correo no tiene un formato válido.'], field: 'email' }
    }
    return { isValid: true, errors: [], field: 'email' }
  }
}
